python3Packages.tree-sitter-grammars: init at 0.22.5

sarcasticadmin opened the pull request:
## Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->
As request in https://github.com/ngi-nix/projects/issues/2104

This generates a python binding package for each tree-sitter-grammar that exists in `tree-sitter.builtGrammars`. 

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

sarcasticadmin committed:
maintainers: add adfaure

sarcasticadmin commented on pkgs/development/python-modules/tree-sitter-grammars/default.nix:
```diff
@@ -0,0 +1,140 @@
+      };
+    };
+in
+# TODO pkgset or flattened?
```
We didn't see any pkgsets inside python3Packages so we will go ahead and flatten this unless we're told otherwise. But we didn't any flattening in top-level either so we guess this will be the first.

mightyiam replied:
We changed our minds. We saw at least two package sets nested in `python3Packages`. We saw `bootstrap` and `qt6`. And this one is literally mapped from an existing package set (`tree-sitter.builtGrammars`) so it seems appropriate.

mightyiam commented on pkgs/top-level/python-packages.nix:
```diff
@@ -15483,6 +15483,17 @@ self: super: with self; {
+  tree-sitter-grammars = lib.recurseIntoAttrs (lib.mapAttrs
+    (name: grammarDrv: callPackage ../development/python-modules/tree-sitter-grammars { inherit name grammarDrv; })
+    (lib.attrsets.removeAttrs pkgs.tree-sitter.builtGrammars [
+      # ImportError: /nix/store/7w7piy6hpdj1swdg5r0bz47gk2g3q855-tree-sitter-perl-grammar-0.22.5/parser: undefined symbol: isnumber
+      "tree-sitter-perl"
+      # ImportError: /nix/store/6rkhjf2mhwnfqwq2962fvv9bnd4xmpk7-python3.11-python-tree-sitter-ql-dbscheme-0.22.5/lib/python3.11/site-packages/tree_sitter_ql_dbscheme/_binding.abi3.so: undefined symbol: tree_sitter_ql_dbscheme
+      "tree-sitter-ql-dbscheme"
+      # ImportError: /nix/store/ybmnykhrgj6sghw4r7ypf897c4wrncr5-python3.11-python-tree-sitter-org-nvim-0.22.5/lib/python3.11/site-packages/tree_sitter_org_nvim/_binding.abi3.so: undefined symbol: tree_sitter_org_nvim
+      "tree-sitter-org-nvim"
```
Each of these might be resolved by simply bumping its source.

Some could possibly be resolved with some tinkering, which we decided should not delay the submitting of this PR for review.

mightyiam replied:
Resolved. All grammars pass!

mightyiam commented:
For the record, this is already ready for review, even though it's marked as draft.

mightyiam commented on pkgs/top-level/python-packages.nix:
```diff
@@ -15486,6 +15486,13 @@ self: super: with self; {
+  tree-sitter-grammars = lib.recurseIntoAttrs (lib.mapAttrs
+    (name: grammarDrv: callPackage ../development/python-modules/tree-sitter-grammars { inherit name grammarDrv; })
+    (lib.attrsets.removeAttrs pkgs.tree-sitter.builtGrammars [
+      # ImportError: /nix/store/7w7piy6hpdj1swdg5r0bz47gk2g3q855-tree-sitter-perl-grammar-0.22.5/parser: undefined symbol: isnumber
+      "tree-sitter-perl"
```
May be blocked by https://github.com/ganezdragon/tree-sitter-perl/issues/45.

mightyiam replied:
This is resolved.

mightyiam commented:
The final commit message and PR title should be updated.

Janik-Haag commented on pkgs/development/python-modules/tree-sitter-grammars/default.nix:
```diff
@@ -0,0 +1,128 @@
+                  ["-std=c11"] if system() != 'Windows' else []
+                ),
+                define_macros=[
+                  ("Py_LIMITED_API", "0x03080000"),
```
What does the magic number/address `0x03080000` here mean, could you please add a comment?

stepbrobd replied:
`0x03080000` defines the lowest Python version our extension supports to be 3.8

Official docs: <https://docs.python.org/3/c-api/stable.html#c.Py_LIMITED_API>

The usage here follows upstream examples:
- https://github.com/tree-sitter/tree-sitter-ocaml/blob/0b12614ded3ec7ed7ab7933a9ba4f695ba4c342e/setup.py#L49
- https://github.com/tree-sitter/tree-sitter-typescript/blob/4f3eb6655a1cd1a1f87ef10201f8e22886dcd76e/setup.py#L49
- https://github.com/tree-sitter/tree-sitter-ruby/blob/0ffe457fb6aabf064f173fd30ea356845cef2513/setup.py#L48

Janik-Haag replied:
That should be documented, and based on that you should also set `disabled = pythonOlder "3.8"`

Janik-Haag commented on pkgs/development/python-modules/tree-sitter-grammars/default.nix:
```diff
@@ -0,0 +1,128 @@
+  inherit version;
+  pname = drvPrefix;
+
+  src = symlinkJoin {
+    name = "${drvPrefix}-source";
+    paths = [
+      (writeTextDir "${snakeCaseName}/__init__.py"
```
To be entirely frank, this seems like madness. It's hard to maintain code, just inline in some file, and I'm fairly certain you could have accomplished the same result using https://github.com/grantjenks/py-tree-sitter-languages (which is already packaged) and by putting less than 15-mins of work into https://github.com/NixOS/nixpkgs/blob/nixos-unstable/pkgs/development/python-modules/tree-sitter/default.nix so it has a package option for additional grammars.

stepbrobd replied:
[grantjenks/py-tree-sitter-languages](https://github.com/grantjenks/py-tree-sitter-languages) is using a deprecated API:

- code: https://github.com/grantjenks/py-tree-sitter-languages/blob/42f4baffec92848be4937b0cc52b2872201fe322/tree_sitter_languages/core.pyx#L14
- deprecation notice: https://github.com/tree-sitter/py-tree-sitter/tree/v0.21.3#build-from-source
- relevant discussion: https://discourse.nixos.org/t/need-help-enabling-grammars-in-treesitter-python/39500/7

Janik-Haag reviewed (changes requested):
Did you consider taking a look at https://github.com/tree-sitter/py-tree-sitter before doing all the work here?

Taking a look at https://github.com/ngi-nix/ngipkgs/issues/139 I'm honestly a bit confused by the `shell.nix` part of the request.
Because you can just do something like:
```nix
let
  pkgs = import <nixpkgs> { };
  tree-sitter-with-grammers = pkgs.tree-sitter.override {
    extraGrammars = with pkgs.tree-sitter-grammars; [
      # put your grammars here
      tree-sitter-go
      tree-sitter-yang
    ];
  };
in pkgs.mkShell {
  buildInputs = [
    tree-sitter-with-grammers
  ];
}
```
And get a shell with tree-sitter and the grammars you want, but I might be missing something.

Regarding [armijnhemel/proximity_matcher_webservice](https://github.com/armijnhemel/proximity_matcher_webservice/) and the extra context in https://discourse.nixos.org/t/need-help-enabling-grammars-in-treesitter-python/39500 from just googeling a bit, it seems like there is official tree-sitter python bindings https://github.com/tree-sitter/py-tree-sitter/ and we already have them packaged in nixpkgs.
It appears like our package needs a little additional work so you can configure grammars, but that should be an easy <15-min job since you just need to add the grammar python wheel to the pythonPath. (at least that's how I read the [language section in the python binding readme](https://github.com/tree-sitter/py-tree-sitter/?tab=readme-ov-file#install-languages)

---

Additonaly please provide reasoning and considerations in your commit message, esp. if you want to add a bunch of hard to maintain code like this. And I'm pretty certain that we should not go with this solution, but instead use the offical python bindings, that are already packaged and a lot easier to maintain.

adfaure commented on pkgs/development/python-modules/tree-sitter-grammars/default.nix:
```diff
@@ -0,0 +1,128 @@
+    ];
+  };
+
+  preCheck = ''
+    rm -r ${snakeCaseName}
+  '';
```
This is "common" for python packages, it seems that `pytest` gets confused if a directory with the same name as the module is present during tests, (in this case `${snakeCaseName}` corresponds to the module name provided by the *generated* grammar package).

It looks like it is fairly common for python packages (here an example: https://github.com/NixOS/nixpkgs/blob/nixos-unstable/pkgs/development/python-modules/tree-sitter/default.nix#L45).

Janik-Haag replied:
Yes, and if you take a look at the file you just linked, you will notice that it has a comment linking to the issue in question on why you have to remove this directory, so please also add a comment here.

doronbehar replied:
The correct comment would be to mention https://github.com/NixOS/nixpkgs/issues/255262 .

stepbrobd commented on pkgs/development/python-modules/tree-sitter-grammars/default.nix:
```diff
@@ -0,0 +1,128 @@
+let
+  inherit (grammarDrv) version;
+
+  snakeCaseName = lib.replaceStrings [ "-" ] [ "_" ] name;
+  drvPrefix = "python-${name}";
+  langIdentOverrides = {
+    tree_sitter_org_nvim = "tree_sitter_org";
+  };
+  langIdent = langIdentOverrides.${snakeCaseName} or snakeCaseName;
```
Personally I believe this is cleaner and easier to maintain than a if statement

What if new a tree-sitter grammar gets added with a weird name like `tree_sitter_org_nvim` 😂? We can simply put another "override" in there

adfaure replied:
Hi, back at this issue.

The issue you pointed out is about avoiding name collisions if multiple grammars end up sharing the same name (`tree_sitter_org`) in nixpkgs.

The renaming in our package is related to how the shared objects are compiled, regardless of the grammar names in Nix. Specifically, the grammar is compiled with the name `tree_sitter_org`, and the binding relies on this symbol name when loading the grammar.

Without this renaming, the binding fails to work. The test results in the following error:
```
       > Traceback (most recent call last):
       >   File "<string>", line 1, in <module>
       >   File "<string>", line 1, in <lambda>
       >   File "/nix/store/px2nj16i5gc3d4mnw5l1nclfdxhry61p-python3-3.12.7/lib/python3.12/importlib/__init__.py", line 90, in import_module
       >     return _bootstrap._gcd_import(name[level:], package, level)
       >            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
       >   File "<frozen importlib._bootstrap>", line 1387, in _gcd_import
       >   File "<frozen importlib._bootstrap>", line 1360, in _find_and_load
       >   File "<frozen importlib._bootstrap>", line 1331, in _find_and_load_unlocked
       >   File "<frozen importlib._bootstrap>", line 935, in _load_unlocked
       >   File "<frozen importlib._bootstrap_external>", line 995, in exec_module
       >   File "<frozen importlib._bootstrap>", line 488, in _call_with_frames_removed
       >   File "/nix/store/18v5mwkikqr35mar7pfsdad631g1zjf8-python3.12-python-tree-sitter-org-nvim-0.24.3/lib/python3.12/site-packages/tree_sitter_org_nvim/__init__.py", line 1, in <module>
       >     from ._binding import language
       > ImportError: /nix/store/18v5mwkikqr35mar7pfsdad631g1zjf8-python3.12-python-tree-sitter-org-nvim-0.24.3/lib/python3.12/site-packages/tree_sitter_org_nvim/_binding.abi3.so: undefined symbol: tree_sitter_org_nvim
```

stepbrobd commented on pkgs/development/python-modules/tree-sitter-grammars/default.nix:
```diff
@@ -0,0 +1,128 @@
+          from tree_sitter import Language, Parser
+
+
+          def test_language():
+            lang = Language(language())
+            assert lang is not None
+            parser = Parser()
+            parser.language = lang
+            tree = parser.parse(bytes("", "utf-8"))
+            assert tree is not None
+        ''
```
IIRC in one of our mob sessions, someone pointed out tree-sitter parsing cannot fail even if provided code doesn't follow the grammar, and no exception will be thrown no matter what (I'm not 100% sure tho)

I agree that asserting `not None` might not be enough, but I think the best we can do here is to change the assert `not None` to the resulting parse tree is an instance of some type that [py-tree-sitter](https://github.com/tree-sitter/py-tree-sitter) provides

stepbrobd commented:
> Did you consider taking a look at https://github.com/tree-sitter/py-tree-sitter before doing all the work here?

Yep, that's the first thing we've done and we are using it in the [`checkPhase`](https://github.com/NixOS/nixpkgs/blob/e5dc90067193c4befcec75d6f40ead37a84d73c1/pkgs/development/python-modules/tree-sitter-grammars/default.nix#L99-L121)


> Taking a look at [ngi-nix/ngipkgs#139](https://github.com/ngi-nix/ngipkgs/issues/139) I'm honestly a bit confused by the `shell.nix` part of the request. Because you can just do something like:
> ...
> And get a shell with tree-sitter and the grammars you want, but I might be missing something.

This is right for exposing `tree-sitter` and `tree-sitter-<lang>` (grammars) to dev shells.
What [ngi-nix/ngipkgs#139](https://github.com/ngi-nix/ngipkgs/issues/139) wants, is exposing **Python bindings for multiple grammars** to dev shells ([original request](https://discourse.nixos.org/t/need-help-enabling-grammars-in-treesitter-python/39500))


> Regarding [armijnhemel/proximity_matcher_webservice](https://github.com/armijnhemel/proximity_matcher_webservice/) and the extra context in https://discourse.nixos.org/t/need-help-enabling-grammars-in-treesitter-python/39500 from just googeling a bit, it seems like there is official tree-sitter python bindings https://github.com/tree-sitter/py-tree-sitter/ and we already have them packaged in nixpkgs.

`py-tree-sitter` is indeed in nixpkgs, but you'll need to load a grammar to it before you can parse anything ([usage](https://github.com/tree-sitter/py-tree-sitter?tab=readme-ov-file#install-languages))
What this PR adds, is to give `py-tree-sitter` in nixpkgs the ability to **load any/all grammars** exposed through `tree-sitter.builtGrammars` as python bindings
Please note that Python bindings exposed through official tree-sitter org (it's mostly distributed through one of the maintainers' personal Pypi account) on Pypi is [incomplete](https://pypi.org/user/amaanq/), 24 as of me writing this. With our way, we can expose 100+ bindings readily available in nixpkgs

> It appears like our package needs a little additional work so you can configure grammars, but that should be an easy <15-min job since you just need to add the grammar python wheel to the pythonPath. (at least that's how I read the [language section in the python binding readme](https://github.com/tree-sitter/py-tree-sitter/?tab=readme-ov-file#install-languages)

That's what we did in this PR: giving users the ability to get tree-sitter grammar python bindings, which will further allow `py-tree-sitter` to load the bindings and be able to parse code

Atemu commented on pkgs/development/tools/parsing/tree-sitter/grammars/tree-sitter-make.json:
```diff
@@ -1,7 +1,7 @@
   "url": "https://github.com/alemuller/tree-sitter-make",
   "rev": "a4b9187417d6be349ee5fd4b6e77b4172c6827dd",
-  "date": "2021-12-16T17:14:17+00:00",
+  "date": "2021-12-16T17:14:17Z",
```
ISO 8601 gives you the option to use either `Z` or `+00:00`, so both are the same timestamp but it's still odd that it changed.

Janik-Haag replied:
Thanks I didn't know about that :D 
In that case, we should take a look in the script and try to check what changed the behavior to avoid

stepbrobd replied:
How about we allow this to happen here, and see what happens with the next update?

adfaure replied:
Hi, looking at this. 

The date is generated from this [line](https://github.com/NixOS/nixpkgs/blob/7f2dd685928306ea256657c56043d3af6396d9e7/pkgs/build-support/fetchgit/nix-prefetch-git#L325C4-L325C75) which is used in the update script to get data.

However, it seems to be caused by the version of git:

```bash
$ git clone git@github.com:alemuller/tree-sitter-make.git
$ cd tree-sitter-make

$ nix-shell -I nixpkgs=https://github.com/NixOS/nixpkgs/archive/24.05.tar.gz -p git
$ git --version
> git version 2.44.1
$ git show -1 --no-patch --pretty=%cI a4b9187
> 2021-12-16T17:14:17+00:00

$ nix-shell -I nixpkgs=https://github.com/NixOS/nixpkgs/archive/24.11.tar.gz -p git
$ git --version
> git version 2.47.0
$ git show -1 --no-patch --pretty=%cI a4b9187
2021-12-16T17:14:17Z 
```

mightyiam commented on pkgs/development/python-modules/tree-sitter-grammars/default.nix:
```diff
@@ -0,0 +1,128 @@
+  };
+  langIdent = langIdentOverrides.${snakeCaseName} or snakeCaseName;
+in
+assert (lib.assertMsg (!(grammarDrv.meta ? license)) ''
+  As of this writing, ${grammarDrv.pname} surprisingly doesn't have a license.
+  This trap is set here to guarantee that if it ever does have a license, this package will inherit the license.
+'');
```
> Please remove this nonsensical assert statement.

Please no name-calling my work.

stepbrobd replied:
We've set all the licenses to MIT

Janik-Haag reviewed (dismissed):
>  What ngi-nix/ngipkgs#139 wants, is exposing Python bindings for multiple grammars to dev shells (original request)

As I said, I wasn't sure if it meant just tree-sitter, or if it meant python bindings.

> What this PR adds, is to give py-tree-sitter in nixpkgs the ability to load any/all grammars exposed through tree-sitter.builtGrammars as python bindings

No, it does not, I think? You aren't touching the py-tree-sitter package, nor are you using py-tree-sitter in your code.

> py-tree-sitter is indeed in nixpkgs, but you'll need to load a grammar to it before you can parse anything (usage)

Yes, that's the point I made, and you aren't adding support for it here.

All you would have needed to do is a small patch like:
```diff
From 12e830ed9a7bace1ef0df64ac088b0a4b309d8cd Mon Sep 17 00:00:00 2001
From: "Janik H." <janik@aq0.de>
Date: Wed, 26 Jun 2024 13:28:58 +0200
Subject: [PATCH] python3Packages.tree-sitter: add support for additional
 grammars

---
 pkgs/development/python-modules/tree-sitter/default.nix | 7 +++++--
 1 file changed, 5 insertions(+), 2 deletions(-)

diff --git a/pkgs/development/python-modules/tree-sitter/default.nix b/pkgs/development/python-modules/tree-sitter/default.nix
index fdaa03554433..0127b1518fa1 100644
--- a/pkgs/development/python-modules/tree-sitter/default.nix
+++ b/pkgs/development/python-modules/tree-sitter/default.nix
@@ -5,11 +5,12 @@
   pytestCheckHook,
   pythonOlder,
   setuptools,
-  tree-sitter-python,
-  tree-sitter-rust,
   tree-sitter-html,
   tree-sitter-javascript,
   tree-sitter-json,
+  tree-sitter-python,
+  tree-sitter-rust,
+  extraGrammars ? [ ]
 }:
 
 buildPythonPackage rec {
@@ -29,6 +30,8 @@ buildPythonPackage rec {
 
   build-system = [ setuptools ];
 
+  dependencies = extraGrammars;
+
   nativeCheckInputs = [
     pytestCheckHook
     tree-sitter-python
-- 
2.45.1
```
(I'm not sure if the above example patch is the *best* way to do this, you might also want to craft your own python environment in `passthru` to avoid unnecessary recompiles (similar to what we are doing in [octodns](https://github.com/NixOS/nixpkgs/blob/nixos-unstable/pkgs/tools/networking/octodns/default.nix#L50)) 

Example usage would be:
```nix
let
  pkgs = import ./. { };
in pkgs.writers.writePython3Bin "tree-sitter-test.py" {
  libraries = [
    (pkgs.python3Packages.tree-sitter.override { extraGrammars = [
        pkgs.python3Packages.tree-sitter-javascript
      ];
    })
  ];
} ''
import tree_sitter_javascript as tsjavascript
from tree_sitter import Language, Parser

JS_LANGUAGE = Language(tsjavascript.language())
parser = Parser(JS_LANGUAGE)
tree = parser.parse(
  bytes("console.log('Hello World!');", "utf8")
)

cursor = tree.walk()
cursor.goto_first_child()
print(cursor.node.type)
''
```

> Please note that Python bindings exposed through official tree-sitter org (it's mostly distributed through one of the maintainers' personal Pypi account) on Pypi is incomplete, 24 as of me writing this. With our way, we can expose 100+ bindings readily available in nixpkgs

How can you be certain that your bindings work, you barely do any testing, there is a bunch of work going into this upstream to do this properly. And you are basically duplicating out of tree efforts like https://github.com/grantjenks/py-tree-sitter-languages, but with worse support and worse testing.
You can also just use `pkgs.python3Packages.tree-sitter-languages` instead of `pkgs.python3Packages.tree-sitter` in the example above (and replace the language and parser definition) and get access to *all* the languages.
We also don't need to depend on the wheels distributed through pypi, for example the tree-sitter-javascript package listed above, was built from source.

Janik-Haag commented:
cc @doronbehar since you did a lot of work on tree-sitter python bindings in #316901 recently.

doronbehar commented:
I'm sorry but this PR is very confusing to me, and I tend to agree with most of @Janik-Haag's comments, and for sure using `writeTextDir` like that is way too messy.

I'm mostly confused because I didn't find any details in neither [the discourse thread](https://discourse.nixos.org/t/need-help-enabling-grammars-in-treesitter-python/39500), nor at https://github.com/ngi-nix/ngipkgs/issues/139 and neither here about the target development setup. For example I wonder:

- What text editor do you use?
- Many text editors support installing such Grammers via the editor's package manager, why is that not an option? In the case of Neovim for example, we have an interface for that for sure ([example usage](https://gitlab.com/doronbehar/nixos-configs/-/blob/da0d0732683dc3871ee6ee8778a06faed876b7a3/overlays/shell-programs.nix#L39-76)). It is purely documented but still it is there.

I'm not familiar with the ecosystem of all of these python packages, including many of those I added in https://github.com/NixOS/nixpkgs/pull/316901 , so I won't be able to help a lot.

jtojnar commented on pkgs/development/python-modules/tree-sitter-grammars/default.nix:
```diff
@@ -0,0 +1,128 @@
+          from setuptools import Extension, setup
+
+
+          setup(
```
Most of this should be moved to `pyproject.toml`, see https://setuptools.pypa.io/en/latest/userguide/ext_modules.html#building-extension-modules

adfaure replied:
Hi, thanks for the suggestion. Indeed, the file is present upstream and I added it.

However, since everything is generated, I am not sure if it is really helpful for us now?

jtojnar commented on pkgs/development/python-modules/tree-sitter-grammars/default.nix:
```diff
@@ -0,0 +1,128 @@
+  inherit version;
+  pname = drvPrefix;
+
+  src = symlinkJoin {
```
It would probably be cleaner keeping the files in-tree, pointing `src` there and using `substituteInPlace --subst-var-by` in `postPatch` to pass the variables.

adfaure replied:
Hi, that is indeed a solution. However, note that two files are generated from the variable: ${snakeCaseName}/__init__.py and ${snakeCaseName}/binding.c.

This may require some file movement. Perhaps a hybrid solution could work as well.

jtojnar commented:
I would say this is pretty much what I imagined in https://discourse.nixos.org/t/need-help-enabling-grammars-in-treesitter-python/39500/7?u=jtojnar

> No, it does not, I think? You aren't touching the py-tree-sitter package, nor are you using py-tree-sitter in your code.

The grammars have nothing to do with `py-tree-sitter` package – the package just contains bindings to `tree-sitter` library itself, I am pretty sure.

And as you show in your example – you just pull in `tree_sitter_javascript` from `PYTHONPATH`. So I do not think the `tree_sitter` bindings module does need to be aware of the grammars at build time, passing the result of `language()` function to `Language` should be enough.

> Yes, that's the point I made, and you aren't adding support for it here.

Again, no support is necessary in `py-tree-sitter`, we just need the grammar bindings, which is what this PR provides as mentioned [above](https://github.com/NixOS/nixpkgs/pull/320783#issuecomment-2190499252). It is creating them from scratch because most grammars are not actually available on PyPI. Also those that are available there are using vendored grammars and vendoring is generally undesirable.

> All you would have needed to do is a small patch like:

I do not think this will do anything that could not be achieved by just adding the `extraGrammars` on Python path in the consumer. The grammars mentioned in the `py-tree-sitter` expression are only used for tests.

> How can you be certain that your bindings work, you barely do any testing, there is a bunch of work going into this upstream to do this properly. And you are basically duplicating out of tree efforts like https://github.com/grantjenks/py-tree-sitter-languages, but with worse support and worse testing.

Yes, testing is an issue

> You can also just use `pkgs.python3Packages.tree-sitter-languages` instead of `pkgs.python3Packages.tree-sitter` in the example above (and replace the language and parser definition) and get access to _all_ the languages.

but tree-sitter-languages is a dead end as I mentioned in https://discourse.nixos.org/t/need-help-enabling-grammars-in-treesitter-python/39500/7?u=jtojnar

The only other alternative to have all those grammars available from Python with the current API (to my knowledge, as of 2024-04-12) would be going around and opening PRs to add a trivial binding like this https://github.com/tree-sitter/tree-sitter-python/tree/71778c2a472ed00a64abf4219544edbf8e4b86d7/bindings/python/tree_sitter_python for every upstream grammar, having them upload it to PyPI, and then packaging that. That does not scale.

> We also don't need to depend on the wheels distributed through pypi, for example the tree-sitter-javascript package listed above, was built from source.

That would be fine but it still does not solve the issue for grammars that do not have Python bindings.

I would say the requirement to have m×n bindings for each combination of language and grammar itself is something that should be addressed but I do not have a solution for that 🤷‍♀️

adfaure commented:
> The only other alternative to have all those grammars available from Python with the current API (to my knowledge, as of 2024-04-12) would be going around and opening PRs to add a trivial binding like this https://github.com/tree-sitter/tree-sitter-python/tree/71778c2a472ed00a64abf4219544edbf8e4b86d7/bindings/python/tree_sitter_python for every upstream grammar, having them upload it to PyPI, and then packaging that. That does not scale.

Hi, I just want to mention that every recent enough grammar should have the binding, since the official template contains it: https://github.com/tree-sitter/tree-sitter-embedded-template.

So maybe, all upstream (maintained) bindings should at some point contain the binding.

In this respect, building recent grammars could be done directly by inheriting the source from the grammars list.

jtojnar commented:
@adfaure That is actually a grammar for [ERB](https://ruby-doc.org/stdlib-2.5.1/libdoc/erb/rdoc/ERB.html) and [EJS](http://ejs.co/) template languages. But looks like a [template does exist](https://github.com/tree-sitter/tree-sitter/tree/9d1ac2139221b2e19389ca620767c537607a2f63/cli/src/generate/templates) so you might be right that it will be everywhere eventually.

So maybe we could use the upstream sources. For example, here is how I would convert `python3.pkgs.tree-sitter-javascript` (after applying #324161):

```patch
--- a/pkgs/development/python-modules/tree-sitter-javascript/default.nix
+++ b/pkgs/development/python-modules/tree-sitter-javascript/default.nix
@@ -1,6 +1,6 @@
 { lib
 , buildPythonPackage
-, fetchFromGitHub
+, tree-sitter-grammars
 , setuptools
 , wheel
 , tree-sitter
@@ -8,16 +8,9 @@
 
 buildPythonPackage rec {
   pname = "tree-sitter-javascript";
-  version = "0.21.3";
+  inherit (tree-sitter-grammars.tree-sitter-javascript) version src;
   pyproject = true;
 
-  src = fetchFromGitHub {
-    owner = "tree-sitter";
-    repo = "tree-sitter-javascript";
-    rev = "v${version}";
-    hash = "sha256-jsdY9Pd9WqZuBYtk088mx1bRQadC6D2/tGGVY+ZZ0J4=";
-  };
-
   build-system = [
     setuptools
     wheel
```

adfaure commented:
> @adfaure That is actually a grammar for [ERB](https://ruby-doc.org/stdlib-2.5.1/libdoc/erb/rdoc/ERB.html) and [EJS](http://ejs.co/) template languages. But looks like a [template does exist](https://github.com/tree-sitter/tree-sitter/tree/9d1ac2139221b2e19389ca620767c537607a2f63/cli/src/generate/templates) so you might be right that it will be everywhere eventually.

Oh, I see. Thank you for pointing that out and finding the actual template.

sarcasticadmin committed:
python3Packages.tree-sitter-grammars: init at 0.22.5

Co-authored-by: Robert James Hernandez <rob@sarcasticadmin.com>
Co-authored-by: Yifei Sun <ysun@hey.com>
Co-authored-by: Ali Jamadi <jamadi1377@gmail.com>
Co-authored-by: yakampe <yanis.kampe.cv@gmail.com>
Co-authored-by: GetPsyched <priyanshu@getpsyched.dev>
Co-authored-by: Adrien Faure <adrien.faure@protonmail.com>
Co-authored-by: Shahar "Dawn" Or <mightyiampresence@gmail.com>

python3Packages.tree-sitter-grammars: Filtering uncompatible grammars

py-tree-sitter is still not up to date to latest tree-sitter (0.25).
py-tree-sitter gets tree-sitter from a submodule which is at version 0.24.

Therefore all grammar with a langage version greater that 14 are not
compatible (language version is most of the time commited into the
source of the grammar, so we don't have a lot of control over it).

For now it concerns two grammars:
- sql grammar gets generated during the build and therefore ends up with
the latest language version (15) not compatible with py-tree-sitter
- templ grammar is already at version 15 in the upstream source.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/2024-summer-of-nix-program-updates/46053/3

mightyiam commented:
I could pick this up again and run it through to completion.

adfaure commented:
Hello everyone, let's wrap this up.

**TLDR:**
- You can build all with: `NIXPKGS_ALLOW_BROKEN=1 NIXPKGS_ALLOW_BROKEN=1 nix-build -A python3Packages.tree-sitter-grammars -k.
`
- I've addressed most of the straightforward comments (e.g., adding comments and minor fixes).
- The generated code may look messy, but it's the best way to make all grammars in nixpkgs work.
- I agree testing is an issue; the provided tests give some information (e.g., it failed for `tree_sitter_nvim_org`).
- I've tried different approaches, and the best solution seems to be enabling tests directly in the `tree-sitter-grammars`.

**Generated Code:**  
With recent versions of tree-sitter, bindings are generated and incorporated into the project. For newer grammars, we could build bindings directly from the source without generating extra files. However, the tree-sitter ecosystem is sparse, and many grammars aren't up-to-date, so they lack bindings. A hybrid solution—generate bindings for old grammars and use direct source for newer ones—might work, but the time-to-benefit ratio might not justify it.

**Testing:**  
Testing is challenging because each grammar defines its own node identifiers in the parsed source tree, making it difficult to have a generic test. The default test checks if the binding can load a language, and tree-sitter has its own testing framework, but it's in Rust and compares code snippets to s-expressions.

I explored three approaches to incorporate tests into the Python modules:

1. **Enable `tree-sitter test` in the `grammar.nix` file (The one that I choose):**
    - This approach revealed broken grammars, which I flagged as broken.
    - I updated the Python script to test the grammar and automatically add the broken flag if the test fails.
2. **Use grammar tests as queries and check them against the root tree:**
    - This worked somewhat, but it's unreliable because the test language (s-expression-based) isn't the same as the query language, and some queries didn't work.
3. **Parse the test language and compare the Python tree with the s-expression:**
    - This approach could work but is too complex and not reliable enough. The test language has many features to implement, and sometimes the generated Python s-expression didn't match the expected s-expression, and I couldn't determine why.

**Usage**

To use the grammars in python, you can install them with (example with rust):

`NIX_PATH=nixpkgs=https://github.com/ngi-nix/nixpkgs/archive/treesitter-python-generated-grammars.tar.gz nix-shell -p 'python3.withPackages(ps : with ps; [ tree-sitter tree-sitter-grammars.tree-sitter-rust ])' --pure`

and execute the following python file `python main.py`:
```python
# main.py
import tree_sitter, tree_sitter_rust

rust = tree_sitter.Language(tree_sitter_rust.language())
parser = tree_sitter.Parser(rust)

tree = parser.parse(
    bytes(
        """
mod english;

mod english {}

mod english {
    mod greetings {}
    mod farewells {}
}

pub mod english;
        """,
        "utf8"
    )
)
```

That should output something like:

```
(source_file (mod_item name: (identifier)) (mod_item name: (identifier) body: (declaration_list)) (mod_item name: (identifier) body: (declaration_list (mod_item name: (identifier) body: (declaration_list)) (mod_item name: (identifier) body: (declaration_list)))) (mod_item (visibility_modifier) name: (identifier)))

fricklerhandwerk commented on pkgs/development/tools/parsing/tree-sitter/update_impl.py:
```diff
@@ -125,12 +154,19 @@ def fetchRepo() -> None:
+            # Load the data since we need the grammar path to execute the tests
+            data = json.loads(res)
+
+            # Test the grammar with Tree-sitter and set the isBroken flag accordingly
+            data['isBroken'] = not check_grammar(data)
```
We should probably only add that if it applies, and remove it if not. This PR now changes lots of files for no good reason.

adfaure committed:
maintainers: add a-jay98

fricklerhandwerk reviewed (commented):
Please add the usage example from https://github.com/NixOS/nixpkgs/pull/320783#issuecomment-2514901215 (with a properly written out shell.nix) to the Nixpkgs manual.

adfaure commented on pkgs/development/python-modules/tree-sitter-grammars/default.nix:
```diff
@@ -0,0 +1,162 @@
+
+  preCheck = ''
+    # https://github.com/NixOS/nixpkgs/issues/255262
+    rm -r ${snakeCaseName}
```
The pytestCheckHook uses the command python -m pytest, and as explained in [this comment](https://github.com/NixOS/nixpkgs/issues/255262), this command adds the current directory to the search path. As a result, we end up executing tests for source files that do not contain the required shared object.

Looking at all the links in issue #255262, there are several ways to address this problem. A few examples include:

    Removing the source files (this is the approach we use).
    Using an alternative test command (e.g., pytest).

fricklerhandwerk replied:
Yeah so "looking at all the links" doesn't scale for readers, but regardless I don't understand the approach: this looks like we're removing the code under test entirely, so what's the point of running the test? Am I missing something?

adfaure replied:
Sorry, I didn't mean for you to go through all the links. I was just pointing out that the solutions I found seem to involve some hacky workarounds for the `pytestCheckHook` behavior.

While I don't think this is directly related to the PR at hand, **I want to better understand the issue so I can explain it more clearly.** I am investigating righ now.

At a high level, my current understanding, it that it is related to how `pytest` loads the tests. We want `pytest` to use the compiled version that contains the shared object, rather than the source version that is not yet ready to be used. And without removing this directory, `pytest` loads the sources.

adfaure replied:
### Summary of the issue
The `pytest -m` command used by the `pytestCheckHook` (`pkgs/development/interpreters/python/hooks/pytest-check-hook.sh`) adds the current directory to `sys.path` (documented [here](https://docs.pytest.org/en/7.1.x/explanation/pythonpath.html#invoking-pytest-versus-python-m-pytest)).

With this setup, the current directory is prioritized before the Python site-packages folder in `sys.path`. Both locations contain a directory named `${snakeCaseName}`. When Python tries to load the module, it prioritizes the current directory, which lacks the compiled binding, causing the test to fail. *The compiled binding resides in `$out/${python.sitePackages}` after the build.*

This issue primarily affects bindings since the source does not include the compiled object.

---

### Example: `tree-sitter-python` grammar (`snakeCaseName` = `tree_sitter_python`)

During testing, the `sys.path` includes the following critical paths (among others). The second and third paths both contain a `tree_sitter_python` directory, but the first path is where the tests reside:
```plaintext
'/build/python-tree-sitter-python-source/tests'
'/build/python-tree-sitter-python-source'
'/nix/store/qs13k1z45jxh1q3hsda470f9r3i0pc86-python3.12-python-tree-sitter-python-0.24.3/lib/python3.12/site-packages'
...
```

---

#### Directory contents:

1. **First directory: `/build/python-tree-sitter-python-source`**
   ```
   .
   ├── build
   │   ├── lib.linux-x86_64-cpython-312
   │   │   └── tree_sitter_python
   │   │       ├── __init__.py
   │   │       ├── _binding.abi3.so
   │   │       └── binding.c
   ...
   ├── tests
   │   └── test_language.py
   ├── tree_sitter_python
   │   ├── __init__.py
   │   └── binding.c
   └── tree_sitter_python.egg-info
   ```

2. **Second directory: `$out/${python.sitePackages}`**
   ```
   ├── tree_sitter_python
   │   ├── __init__.py
   │   ├── _binding.abi3.so
   │   ├── binding.c
   │   └── __pycache__
   └── tree_sitter_python-0.24.3.dist-info
   ```

---

#### Failure mechanism:

When the test imports `tree_sitter_python`, Python looks in the first matching directory in `sys.path`. In this case, it prioritizes `/build/python-tree-sitter-python-source`, which does not contain the compiled `_binding.abi3.so` object. This results in an import error:
```
ImportError: No module named 'tree_sitter_python._binding'
```

---

I hope this raises the confusions.

adfaure commented on pkgs/development/tools/parsing/tree-sitter/update_impl.py:
```diff
@@ -21,6 +22,34 @@ def log(msg: str) -> None:
     print(msg, file=sys.stderr)
 
 
+def copy_grammar_to_tmpdir(path: str, tmpdirname: str):
+    yield "cp"
+    yield "--no-preserve=mode,ownership"
+    yield "-r"
+    yield f"{path}/."
+    yield tmpdirname
+
+
+def check_grammar(data) -> bool:
+    path = data["path"]
+
+    with tempfile.TemporaryDirectory() as tmpdirname:
+        # Copy grammar to a temporary directory
+        run_cmd(copy_grammar_to_tmpdir(path, tmpdirname))
```
Yes, happy to make this change.

adfaure commented on pkgs/development/tools/parsing/tree-sitter/update_impl.py:
```diff
@@ -21,6 +22,34 @@ def log(msg: str) -> None:
+        # Copy grammar to a temporary directory
+        run_cmd(copy_grammar_to_tmpdir(path, tmpdirname))
+
+        # Run the tree-sitter test command
+        res = sub.run(
+            "cd {tmpdirname} ; {tree_sitter_bin} test".format(
+                tmpdirname=tmpdirname, tree_sitter_bin=bins["tree-sitter"]
+            ),
+            stdout=sub.PIPE,
+            shell=True,
+        )
+
+        # Return True if the command succeeded, False otherwise
+        return res.returncode == 0
```
This is clearer, thanks

adfaure commented on pkgs/development/python-modules/tree-sitter-grammars/default.nix:
```diff
@@ -0,0 +1,162 @@
+        build = "cp38-*"
+        build-frontend = "build"
+      '')
+      (writeTextDir "tests/test_language.py" ''
+        from ${snakeCaseName} import language
+        from tree_sitter import Language, Parser
+
+
+        def test_language():
+          lang = Language(language())
+          assert lang is not None
+          parser = Parser()
+          parser.language = lang
+          tree = parser.parse(bytes("", "utf-8"))
+          assert tree is not None
+      '')
```
It caches binding errors, such as when `langIdentOverrides` is ill-defined for a grammar.
I am adding a comment to explain it.

Something like:
```python
        # This test only checks that the binding can load the grammar from the compiled shared object.
        # It does not verify the grammar itself; that is tested in
        # `pkgs/development/tools/parsing/tree-sitter/grammar.nix`.

        def test_language():
          lang = Language(language())
```

adfaure commented on pkgs/development/tools/parsing/tree-sitter/default.nix:
```diff
@@ -60,6 +60,7 @@ let
           src = grammar.src or (fetchGrammar grammar);
           location = grammar.location or null;
           generate = grammar.generate or false;
+          isBroken = builtins.hasAttr "isBroken" grammar && grammar.isBroken || false;
```
Addressed.

adfaure commented on pkgs/development/python-modules/tree-sitter-grammars/default.nix:
```diff
@@ -0,0 +1,162 @@
+      '')
+      (writeTextDir "pyproject.toml" ''
+        [build-system]
+        requires = ["setuptools>=42", "wheel"]
```
I agree, I doubt that we need any version constraint here since we are tied with the one provided by the current version of python we are building for.

adfaure commented on pkgs/development/python-modules/tree-sitter-grammars/default.nix:
```diff
@@ -0,0 +1,162 @@
+  symlinkJoin,
+  writeTextDir,
+  pythonOlder,
+  # `name`: grammar derivation pname in the format of `tree-sitter-<lang>`
+  name,
+  grammarDrv,
```
Do you mean something like this ?

```nix
{
  lib,
  buildPythonPackage,
  pytestCheckHook,
  tree-sitter,
  symlinkJoin,
  writeTextDir,
  pythonOlder,
} : {
  # `name`: grammar derivation pname in the format of `tree-sitter-<lang>`
  name,
  grammarDrv,
} :
```

adfaure commented:
Thank you @fricklerhandwerk for the review, I addressed most of your comments.

I still need to address the following:

> Please add the usage example from [#320783 (comment)](https://github.com/NixOS/nixpkgs/pull/320783#issuecomment-2514901215) (with a properly written out shell.nix) to the Nixpkgs manual.

I'll do that shortly, I just need to locate it first.

> Don't we want to derive that from the Python we use for building? Or if it's fixed to 3.8 for some reason, at least use the same source for it everywhere it's referenced. See all the other hard-coded values.

Regarding Py_LIMITED_API, I don’t have the right answer yet. I need to think about it a bit more.

adfaure committed:
doc/packages/python-tree-sitter: add documentation

fricklerhandwerk commented on maintainers/maintainer-list.nix:
```diff
@@ -850,12 +857,19 @@
     githubId = 4717906;
     name = "Jakub Skokan";
   };
+
```
```suggestion
```

Stray newline

adfaure replied:
Thank you, it should be fixed now.

fricklerhandwerk commented:
I think this is good and will merge it in the next few days. If anyone has complaints after that, we're here to fix forward.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/ngi-nix-report-for-2025q1-2025-01-13-2025-04-15/63859/1
